text/0374-jsii-modular.md

```diff
@@ -0,0 +1,256 @@
+
+For example, this allows the `jsii` compiler to be upgraded to a new release
+line of `typescript` without forcing every downstream consumer to also accept
+the syntax breaking changes at the same time. Instead, consumers will be free
+to take on a new release line of `jsii` independently from their dependencies as
+well as from their consumers.
+
+Separating the mono-repository into multiple repositories will help ensure we
+maintain backwards compatibility between individual packages, and reduce the
+risk of creating codependent packages (different repositories need to be
+released independenlty, so dependencies can only flow in an asyclic manner).
+This will also make it easier for new contributors to approach *jsii*, as they
+will be able to focus strictly on one package, instead of having the entire
+constellation in their working set.
+
+### Why should we _not_ do this?
+
+Breaking up the mono-repository into multiple independent repositories will come
+at a higher maintenance cost than having a single repository. The act of moving
+code around will also require some time to be invested: there are currently
+co-dependent relationships between certain packages (`jsii-pacmak` has a local
+develooment dependency on the runtimes, and on the `jsii-calc` packages, etc...)
+that will need to be broken. Additionally, some checks currently enforce that
+all tools used to work on an assembly are on the same version line, which will
+no longer be the case.
+
+### What is the technical solution (design) of this feature?
+
+The first component of implementing this RFC is to migrate code for various
+tools and libraries to their own repository:
+
+- `@jsii/spec` and `jsii-reflect` (those can be combined as they have similar
+  purpose)
+- `jsii`
+- `jsii-pacmak`
+- `jsii-rosetta`
+- `@jsii/kernel` and `@jsii/runtime` (which could be combined into a single
+  package)
+- Each individual `@jsii/*-runtime` package (and related tests)
+
+Those new repositories could be created within the `aws` GitHub organization, or
+alternatively a new organization could be created (the `jsii` name is already in
+use as a GitHub organization).
+
+The second component of this RFC is changing the release strategy for the `jsii`
+compiler so that it follows `typescript` version lines (`major.minor`), and
+ensuring other downstream consumers of `.jsii` assemblies do not attempt to
+enforce version consistency using the `jsiiVersion` field.
+
+### Is this a breaking change?
+
+This is not a breaking change in the traditional sense: all existing code will
+continue to work with the current release lines of all packages in the
+ecosystem. However, the new versioning strategy of `jsii` is a departure from
+Semantic Versioning, which will be accompanied with a major version change (from
+`1.x` to `4.3.x` - matching the current `typescript` release line). This is to
+be noted in the release ntoes.
+
+### What alternative solutions did you consider?
```

Comment:
This section is discussing alternatives along different axes.

The first paragraph addresses a different "versioning" solution (so an altogether different approach to solving the problem) (fully decoupled versions vs linking major versions to CDK).

The second paragraph addresses, once we've picked fully decoupled versions, an **implementation** concern about how we will version the packages independently (monorepo vs multirepo).

Those concerns do not really belong at the same level. I would suggest making it more clear what aspect of the problem we're targeting when discussing these alternatives.

Reply:
I like to add these as several appendices as "alternatives for <design/choice>" and discuss pros and cons there, and link them off from here.